File: algorithms/066.plus-one.cpp
class Solution {
public:
    vector<int> plusOne(vector<int>& digits) {
        assert(digits.size() > 0);
        bool carry_flag = true;
        int i;
        vector<int> result;
        
        // plus one process with flow digit
        for (i=digits.size() - 1; i >= 0; i--){
            if (digits[i] + 1 == 10)
                digits[i] = 0;
            else{
                digits[i]++;
                carry_flag = false;
                break;
            }
        }
        
        // additional digit 
        if (carry_flag)
            digits.insert(digits.begin(), 1);
            
        return digits;
    }
};

File: algorithms/066.plus-one.go
func reverse(digits []int){
    for i, j := 0, len(digits) - 1; i < j; i, j = i + 1, j - 1 {
        digits[i], digits[j] = digits[j], digits[i]
    }
}

func plusOne(digits []int) []int {
    reverse(digits)
    carry := 1
    for i := 0; i < len(digits) && carry > 0; i++ {
        digits[i], carry = (digits[i] + carry)%10, (digits[i] + carry)/10
    }
    if carry == 1 {
        digits = append(digits, 1)
    }

    reverse(digits)
    return digits
}
